program_transformers/src/rollups/rollup_persister.rs

```diff
@@ -0,0 +1,571 @@
+            Ok(leaf_hash) => leaf_hash,
+            Err(e) => {
+                return Err(e);
+            }
+        };
+        leaf_hashes.push(leaf_hash);
+    }
+
+    merkle_tree_wrapper::validate_change_logs(
+        rollup.max_depth,
+        rollup.max_buffer_size,
+        &leaf_hashes,
+        rollup,
+    )
+}
+
+fn get_leaf_hash(
+    asset: &RolledMintInstruction,
+    tree_id: &Pubkey,
+) -> Result<[u8; 32], RollupValidationError> {
+    let asset_id = get_asset_id(tree_id, asset.leaf_update.nonce());
+    if asset_id != asset.leaf_update.id() {
+        return Err(RollupValidationError::PDACheckFail(
+            asset_id.to_string(),
+            asset.leaf_update.id().to_string(),
+        ));
+    }
+
+    // @dev: seller_fee_basis points is encoded twice so that it can be passed to marketplace
+    // instructions, without passing the entire, un-hashed MetadataArgs struct
+    let metadata_args_hash = keccak::hashv(&[asset.mint_args.try_to_vec()?.as_slice()]);
+    let data_hash = keccak::hashv(&[
+        &metadata_args_hash.to_bytes(),
+        &asset.mint_args.seller_fee_basis_points.to_le_bytes(),
+    ]);
+    if asset.leaf_update.data_hash() != data_hash.to_bytes() {
+        return Err(RollupValidationError::InvalidDataHash(
+            data_hash.to_string(),
+            Hash::new(asset.leaf_update.data_hash().as_slice()).to_string(),
+        ));
+    }
+
+    // Use the metadata auth to check whether we can allow `verified` to be set to true in the
+    // creator Vec.
+    let creator_data = asset
+        .mint_args
+        .creators
+        .iter()
+        .map(|c| [c.address.as_ref(), &[c.verified as u8], &[c.share]].concat())
+        .collect::<Vec<_>>();
+
+    // Calculate creator hash.
+    let creator_hash = keccak::hashv(
+        creator_data
+            .iter()
+            .map(|c| c.as_slice())
+            .collect::<Vec<&[u8]>>()
+            .as_ref(),
+    );
+    if asset.leaf_update.creator_hash() != creator_hash.to_bytes() {
+        return Err(RollupValidationError::InvalidCreatorsHash(
+            creator_hash.to_string(),
+            Hash::new(asset.leaf_update.creator_hash().as_slice()).to_string(),
+        ));
+    }
+
+    Ok(asset.leaf_update.hash())
+}
+
+pub async fn store_rollup_update<T>(
+    slot: u64,
+    signature: String,
+    rollup: &Rollup,
+    txn: &T,
+    cl_audits: bool,
+) -> ProgramTransformerResult<()>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    for rolled_mint in rollup.rolled_mints.iter() {
+        bubblegum::mint_v1::mint_v1(
+            &rolled_mint.into(),
+            &InstructionBundle {
+                txn_id: &signature,
+                program: Default::default(),
+                instruction: None,
+                inner_ix: None,
+                keys: &[],
+                slot,
+            },
+            txn,
+            "CreateTreeWithRoot",
+            cl_audits,
+        )
+        .await?;
+    }
```

Comment:
So what happens if the all these rolled mints go to `mint_v1`, but there's one missing?  Seems like that shouldn't happen because they are fully validated tree down to the merkle root.

But if it was ever possible then I think it results in a missing an entry in the `cl_audits` table, and gets detected by the backfilling process, which would try to get the missing transactions, but they wouldn't exist.  So maybe we want to prevent all rollups from being entered in the `cl_audits` table and just hardcode this `cl_audits` bool to `false`?

But then again the `cl_audits` table might also be used for auditing purposes other than gap filling, so maybe we want to put entries?  Probably needs some additional discussion.

Overall the state management is good so looks like you would have `RollupPersistingState::FailedToPersist` as an indicator that you need to retry or backfill rollups.

Reply:
Hmm
For now i do not think it is possible to miss some asset inside rollup because of validation
And if we will find some setup that create a gap we might just add additional validation check
But for more straightforward flow sake it can be convinient to set cl_audits == false, wdyt?

Reply:
Yes I think it makes sense to set `cl_audits = false` as this will not be part of that backfill system.